wdl-lint/src/rules/nonmatching_output.rs

```diff
@@ -0,0 +1,351 @@
+         workflow, there should be an entry under `meta.outputs` with that same name. \
+         Additionally, these entries should be in the same order (that order is up to the \
+         developer to decide). No extraneous output entries are allowed."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Completeness])
+    }
+}
+
+/// Check each output key exists in the `outputs` key within the `meta` section.
+fn check_matching(state: &mut Diagnostics, rule: &mut NonmatchingOutputRule<'_>) {
+    let mut exact_match = true;
+    // Check for expected entries missing from `meta.outputs`.
+    for (name, span) in &rule.output_keys {
+        if !rule.meta_outputs_keys.contains_key(name) {
+            if exact_match {
+                exact_match = false;
+            }
+            if rule.current_meta_span.is_some() {
+                state.add(nonmatching_output(
+                    *span,
+                    name,
+                    rule.name.as_deref().expect("should have a name"),
+                    rule.ty.expect("should have a type"),
+                ));
+            }
+        }
+    }
+
+    // Check for extra entries in `meta.outputs`.
+    for (name, span) in &rule.meta_outputs_keys {
+        if !rule.output_keys.contains_key(name) {
+            if exact_match {
+                exact_match = false;
+            }
+            if rule.current_output_span.is_some() {
+                state.add(extra_output_in_meta(
+                    *span,
+                    name,
+                    rule.name.as_deref().expect("should have a name"),
+                    rule.ty.expect("should have a type"),
+                ));
+            }
+        }
+    }
+
+    // Check for out-of-order entries.
+    if exact_match && !rule.meta_outputs_keys.keys().eq(rule.output_keys.keys()) {
+        state.add(out_of_order(
+            rule.current_meta_outputs_span
+                .expect("should have a `meta.outputs` span"),
+            rule.current_output_span
+                .expect("should have an `output` span"),
+            rule.name.as_deref().expect("should have a name"),
+            rule.ty.expect("should have a type"),
+        ));
+    }
+}
+
+impl<'a> Visitor for NonmatchingOutputRule<'a> {
+    type State = Diagnostics;
+
+    fn document(&mut self, _: &mut Self::State, reason: VisitReason, _: &Document) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // Reset the visitor upon document entry
+        *self = Default::default();
+    }
+
+    fn workflow_definition(
+        &mut self,
+        state: &mut Self::State,
+        reason: VisitReason,
+        workflow: &WorkflowDefinition,
+    ) {
+        match reason {
+            VisitReason::Enter => {
+                self.name = Some(workflow.name().as_str().to_string());
+                self.ty = Some("workflow");
```

Comment:
Could we follow the example of other lint rules and wrap this in an `enum Context` instead of passing around `str`?

Reply:
We probably could, but that seems a heavy solution to ultimately print `task` or `workflow`. Before going that route, I'd actually prefer if we could store a `TaskOrWorkflow` instance and use that instead, rather than creating a new `enum` to handle it.